src/metrics/dns.rs

```diff
@@ -0,0 +1,278 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::identity::Identity;
+use crate::metrics::{DefaultedUnknown, Recorder};
+use crate::state::workload::address::Address;
+use crate::state::workload::Workload;
+use prometheus_client::encoding::EncodeLabelSet;
+use prometheus_client::metrics::counter::Counter;
+use prometheus_client::metrics::family::Family;
+use prometheus_client::metrics::histogram::Histogram;
+use prometheus_client::registry::Registry;
+use std::time::Duration;
+use trust_dns_server::server::Request;
+
+pub struct Metrics {
+    pub requests: Family<DnsLabels, Counter>,
+    pub forwarded_requests: Family<DnsLabels, Counter>,
+    pub forwarded_failures: Family<DnsLabels, Counter>,
+    pub forwarded_duration: Family<DnsLabels, Histogram>,
+}
+
+impl Metrics {
+    pub fn new(registry: &mut Registry) -> Self {
+        let requests = Family::default();
+        registry.register(
+            "dns_requests_total",
+            "Total number of DNS requests.",
+            requests.clone(),
+        );
+
+        let forwarded_requests = Family::default();
+        registry.register(
+            "dns_upstream_requests_total",
+            "Total number of DNS requests forwarded to upstream.",
+            forwarded_requests.clone(),
+        );
+
+        let forwarded_failures = Family::default();
+        registry.register(
+            "dns_upstream_failures_total",
+            "Total number of DNS requests forwarded to upstream.",
+            forwarded_failures.clone(),
+        );
+
+        let forwarded_duration = Family::<DnsLabels, Histogram>::new_with_constructor(|| {
+            Histogram::new(vec![0.005f64, 0.001, 0.01, 0.1, 1.0, 5.0].into_iter())
+        });
+        registry.register(
+            "dns_upstream_request_duration_seconds",
+            "Total time in seconds Istio takes to get DNS response from upstream.",
+            forwarded_duration.clone(),
+        );
+
+        Self {
+            requests,
+            forwarded_requests,
+            forwarded_failures,
+            forwarded_duration,
+        }
+    }
+}
+
+#[derive(Clone, Hash, Debug, PartialEq, Eq, EncodeLabelSet)]
+pub struct DnsLabels {
+    request_source_ip: String,
+    request_hostname: String,
+    request_query_type: String,
+    request_protocol: String,
+
+    // Source workload.
+    source_workload: DefaultedUnknown<String>,
+    source_canonical_service: DefaultedUnknown<String>,
+    source_canonical_revision: DefaultedUnknown<String>,
+    source_workload_namespace: DefaultedUnknown<String>,
+    source_principal: DefaultedUnknown<Identity>,
+    source_app: DefaultedUnknown<String>,
+    source_version: DefaultedUnknown<String>,
```

Comment:
Largely these are legacy I think, in favor of source_canonical_service/source_canonical_revision. I wonder if we drop them for this new metric